README.md
# bash libs
The purpose is to have files that contain common functionality which can
be sourced and included into other scripts. See bash-pack or makeself to
allow your scripts to be portable.

## Libraries / Files
- arrays - array handling
- lockfile - lockfile handling
- msg-handling - messaging
- os-info - gather OS information
- pass-chooser - determine GPG program/interface
- user-check - determine user access

## Copyright and Thanks
Author: Shawn Wilson <ag4ve.us@gmail.com>
GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007
http://www.gnu.org/licenses/gpl-3.0.html

I thank Korelogic for their support on this project.



ChangeLog
2015-12-13 13:41  swilson

	* lockfile.sh (1.6): Move functions to user-check.sh and call them.

2015-12-13 13:37  swilson

	* user-check.sh (1.6): Add functions to check world-writable
	  directory and writable directory.

2015-12-12 11:09  swilson

	* user-check.sh (1.5): Check EUID as well.

2015-12-12 10:48  hlein

	* lockfile.sh (1.5): Rough implementation of checking that the
	  lockfile destination directory is not world-writable, and is
	  writable by us.

2015-12-12 10:43  swilson

	* user-check.sh (1.4): Correct logic.

2015-12-12 10:41  swilson

	* TODO (1.2), arrays.sh (1.4), lockfile.sh (1.4), msg-handling.sh
	  (1.14), pass-chooser.sh (1.4), user-check.sh (1.3): Add Id lines.

2015-10-29 22:13  swilson

	* msg-handling.sh (1.13): Can not tell if a file is writeable yet.

2015-10-29 21:58  swilson

	* msg-handling.sh (1.12): Define a /private/ function to allow
	  debugging to output to a file.

2015-10-28 06:01  swilson

	* msg-handling.sh (1.11): Correct syntax.

2015-10-27 11:41  swilson

	* msg-handling.sh (1.10): Do not try to do math if debug contains
	  has =.

2015-10-27 11:23  swilson

	* msg-handling.sh (1.9): Fix logic.

2015-10-27 11:20  swilson

	* msg-handling.sh (1.8): Correct documentation.

2015-10-27 11:17  swilson

	* msg-handling.sh (1.7): Treat debug numbers are absolute values
	  and =number just prints that debug value.

2015-10-27 10:53  swilson

	* msg-handling.sh (1.6): Without a level, should not print a
	  message if debug is 0.

2015-10-27 10:11  swilson

	* msg-handling.sh (1.5): Correction.

2015-10-27 10:01  swilson

	* msg-handling.sh (1.4): Style.

2015-10-27 10:00  swilson

	* msg-handling.sh (1.3): Allow signs in numbers and only print
	  absolute debug level messages for negative values and allow for
	  no level in debug.

2015-10-25 21:04  swilson

	* pass-chooser.sh (1.3): Add whatpass resource.

2015-10-25 20:42  swilson

	* pass-chooser.sh (1.2): Add documentation/resources and remove
	  path variable (was not well baked, potential to overwrite the a
	  veriable of the caller, etc).

2015-10-25 20:15  swilson

	* arrays.sh (1.3): Add dependancy.

2015-10-24 16:08  swilson

	* arrays.sh (1.2), lockfile.sh (1.3): Update documentationa.

2015-09-23 07:02  swilson

	* msg-handling.sh (1.2): Match numbers - not everything but.

2015-09-23 06:44  swilson

	* lockfile.sh (1.2), user-check.sh (1.2): Can not source as a ./ as
	  the path is indeterminate when packing them.

2015-09-14 16:08  swilson

	* user-check.sh (1.1): Initial file to check user permissions.

2015-09-09 07:07  swilson

	* TODO (1.1), arrays.sh (1.1), lockfile.sh (1.1), msg-handling.sh
	  (1.1), pass-chooser.sh (1.1): Split out common bash functions
	  into a libs directory.



os-info.sh
#!/bin/bash

#####
# os-info - returns basic os info
# * NOTE: this may run any code in /etc/*-{version,release} and pollute
# your veriable namespace
#####

#####
# os_ver - returns (amont other possible clutter) $OS and $VER
#####

os_ver()
{
  if [[ -f /etc/os-release ]];then
    # freedesktop.org and systemd
    . /etc/os-release
    OS="$NAME"
    VER="$VERSION_ID"
  elif type lsb_release >/dev/null; then
    # linuxbase.org
    OS="$(lsb_release -si)"
    VER="$(lsb_release -sr)"
  elif [[ -f /etc/lsb-release ]]; then
    # For some versions of Debian/Ubuntu without lsb_release command
    . /etc/lsb-release
    OS="$DISTRIB_ID"
    VER="$DISTRIB_RELEASE"
  elif [[ -f /etc/debian_version ]]; then
    # Older Debian/Ubuntu/etc.
    OS=Debian
    VER=$(< /etc/debian_version)
# TODO
#  elif [ -f /etc/SuSe-release ]; then
#  elif [ -f /etc/redhat-release ]; then
  else
    # Fall back to uname, e.g. "Linux <version>", also works for BSD, etc.
    OS="$(uname -s)"
    VER="$(uname -r)"
  fi
}



lockfile.sh
#!/bin/bash -

# $Id: lockfile.sh,v 1.6 2015/12/13 20:41:31 swilson Exp $

#####
# lockfile - functions for lockfile handling
#####

source msg-handling.sh
source user-check.sh

#####
# lockfile <filename>
# Atomically create a lockfile
#####
lockfile()
{
  debug 6 "lockfile [$@]"
  if [[ -z "$1" ]] ; then
    die 2 "No lockfile info defined\n"
  fi
  declare lockfile="$1"
  declare pid

  # Sanity checks
  not_world_write "$lockfile"
  can_write "$lockfile"

  # Check the pid of an old lockfile
  if [[ -f "$lockfile" ]] ; then
    debug 5 "[$lockfile] exists"
    pid=$(< $lockfile)
    if [[ $pid == *[!0-9]* ]]; then
      die 2 "Pid is not a number [$pid]\n"
    fi
    debug 4 "Prior pid [$pid]"
    if kill -0 $pid >/dev/null 2>&1 ; then
      die 2 "Lockfile [$lockfile] exists with pid [$pid]\n"
    else
      # Orphaned
      rm -f "$lockfile"
      debug 2 "Removed stale lock file [$lockfile]"
    fi
  fi

  # Write pid to a temp lockfile
  echo -n "$$" >"$lockfile.$$" 

  # Avoid race - ln will fail if the file is there and create the file if
  # it isn't - should be consistent on any posix filesystem
  if ln "$lockfile.$$" "$lockfile" ; then
    debug 5 "Lock obtained"
    rm -f "$lockfile.$$"
    trap "cleanup 0 $lockfile" EXIT
  else
    rm -f "$lockfile.$$"
    die 2 "Lockfile [$lockfile] exists\n"
  fi

  trap "cleanup 3 $lockfile" 1 2 15
}

#####
# cleanup <status> <file0, file1, file2...>
# Remove a list of files
#####
cleanup()
{
  debug 6 "cleanup [$@]"
  status="$1"
  shift
  for file ; do
    if [[ -f "$file" ]] ; then
      debug 5 "Removing [$file]"
      rm -f "$file"
    else
      warn "Cleanup file does not exist [$file]"
    fi
  done
  die "$status" "Done cleanup\n"
}



msg-handling.sh
#!/bin/bash

# $Id: msg-handling.sh,v 1.14 2015/12/12 17:41:56 swilson Exp $

#####
# msg-handling - contains functions to handle messages.
#####

#####
# die [status] "<message>\n"
#####
die()
{
  if [[ "${1#[-+]}" == [0-9]* ]] ; then
    status="$1"
    shift
  else
    status=1
  fi
  echo -ne "$@" >&2
  exit "$status"
}

#####
# warn <message>
#####
warn()
{
  echo -ne "WARN: $@\n" >&2
}

#####
# debug <level> "<message>"
# print a message for levels up to and including the absolute debug value or
# print message that match the debug levels if =<debug> values is used
# writes to $debug_file if it is set
#####
debug()
{
  : "${debug:=0}"
  declare level
  if [[ "${1#[+-]}" == [0-9]* ]] ; then
    level="${1#[+-]}"
    shift
  fi
  local output="$@"

  # Define a proper output function
  if ! declare -f _debug_out >/dev/null 2>&1 ; then
    if [[ -n "${debug_file:-}" ]] ; then
      _debug_out()
      {
        echo -ne "$@\n" >> "${debug_file:-}"
      }
    else
      _debug_out()
      {
        warn "DEBUG $@"
      }
    fi
  fi
  
  if [[ -n "${level:-}" ]] ; then
    # Pin to a debug level
    if [[ "$debug" == "="* && "${debug#=}" == "$level" ]] ; then
      _debug_out "[$level]: $output"
    # Less than or equal to level
    elif [[ "$debug" != "="* ]] && (( level <= debug )) ; then
      _debug_out "[$level]: $output"
    fi
  # Script didn't set a debug level
  elif (( debug != 0 )) ; then
    _debug_out "[nul]: $output"
  fi
}



pass-chooser.sh
#!/bin/bash -

# $Id: pass-chooser.sh,v 1.4 2015/12/12 17:41:56 swilson Exp $

#####
# choose-pass - determine which program to use to return a passphras
#####

source msg-handling.sh

#####
# passprog
# returns getpass variable
# if GETPASS is set to a valid password manager, tries to use it
#####
passprog()
{
  declare -g getpass
  # http://www.passwordstore.org/
  if type pass >/dev/null 2>&1 &&
    [[ ! -v GETPASS || $GETPASS == "pass" ]] ; 
  then
    getpass="pass"
    debug 2 "Using pass"
  # http://korelogic.com/Resources/Tools/whatpass
  elif type whatpass >/dev/null 2>&1 &&
    [[ ! -v GETPASS || $GETPASS == "whatpass" ]] ; 
  then
    getpass="whatpass"
    debug 2 "Using whatpass"
  # https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/security.1.html
  elif type security >/dev/null 2>&1 &&
    [[ ! -v GETPASS || $GETPASS == "security" ]] ;
  then
    getpass="security"
    debug 2 "Using security"
  # https://www.gnupg.org/
  elif type gpg >/dev/null 2>&1 &&
    [[ ! -v GETPASS || $GETPASS == "gpg" ]] ;
  then
    getpass="gpg"
    debug 2 "Using gpg"
  else
    die "No pass or whatpass."
  fi
}
